### PR TITLE
Fix rounding when very close to an integer

### PR DESCRIPTION
**Pull request by kklobe**
# Description

This is an experimental hack for #2404. When in `ROUND_Chop` mode on non-x64 platforms, if the input number is very close to, but less than, a whole integer, it gets chopped to the value below, e.g. 7.9999999994 will chop to 7 when it should be 8.

This is only done when the FPU is set to extended precision mode, the assumption being that the number to chop, if that close to an integer, is inaccurate and would likely be the integer under full 80-bit precision.

# Manual testing

Tested the options sliders with Quake SW 1.06, and also tested QTD for performance regressions (none noted).

# Checklist

_Please tick the items as you have addressed them. Don't remove items; leave the ones that are not applicable unchecked._

I have:

- [x] followed the project's [contributing guidelines](https://github.com/dosbox-staging/dosbox-staging/blob/master/CONTRIBUTING.md) and [code of conduct](https://github.com/dosbox-staging/dosbox-staging/blob/master/CODE_OF_CONDUCT.md).
- [x] performed a self-review of my code.
- [x] commented on the particularly hard-to-understand areas of my code.
- [x] split my work into well-defined, bisectable commits, and I [named my commits well](https://github.com/dosbox-staging/dosbox-staging/blob/main/CONTRIBUTING.md#commit-messages).
- [x] applied the appropriate labels (bug, enhancement, refactoring, documentation, etc.)
- [x] [checked](https://github.com/dosbox-staging/dosbox-staging/blob/main/scripts/compile_commits.sh) that all my commits can be built.
- [x] confirmed that my code does not cause performance regressions (e.g., by running the Quake benchmark).
- [ ] added unit tests where applicable to prove the correctness of my code and to avoid future regressions.
- [ ] made corresponding changes to the documentation or the website according to the [documentation guidelines](https://github.com/dosbox-staging/dosbox-staging/blob/main/DOCUMENTATION.md).
- [ ] [locally verified](https://github.com/dosbox-staging/dosbox-staging/blob/main/DOCUMENTATION.md#previewing-documentation-changes-locally) my website or documentation changes.

### CODE REVIEW

**Comment by kcgen** [comment]
(Have my system switched over in FreeBSD / PPC land; will help test when I get through it :)

Here's Intel's FPU tester [mcpdiag.zip](https://github.com/dosbox-staging/dosbox-staging/files/13461505/mcpdiag.zip); repro-pack will autotype to launch the test.

**Comment by kklobe** [comment]
> (Have my system switched over in FreeBSD+PPC land; will help test when I get through it :)
> 
> Here's Intel's FPU tester [mcpdiag.zip](https://github.com/dosbox-staging/dosbox-staging/files/13461505/mcpdiag.zip); repro-pack will autotype to launch the test.

Yep - 8 more "Round to Integer" failures on this branch, which would be expected in a strict compliance test.

**Comment by kcgen** [comment]
> Yep - 8 more "Round to Integer" failures on this branch, which would be expected in a strict compliance test.

:+1: ; makes sense - but maybe we're still at-par (or better) on ARM64?

**Comment by kklobe** [comment]
> > Yep - 8 more "Round to Integer" failures on this branch, which would be expected in a strict compliance test.
> 
> 👍 ; makes sense - but maybe we're still at-par (or better) on ARM64?

Yeah - main vs this branch on the M1 has the 8 add'l failures, but I think this hack provides the true intent of code hitting it, like the Quake scenario: if someone is calling FRNDINT with 7.9999999999999998 or 8.0000000000000003, in `ROUND_Chop` mode, they almost certainly were passing in 8.

**Comment by Burrito78** [comment]
M2 MacBook Air

MCPDIAG: Only difference of this PR is that it passes 8 tests less than main in the category "Round to Integer" (as expected).

There is still an issue in Quake though. The slider still doesn't go to 11 err completely to the right. But it's getting better!

main
main

PR
PR

**Comment by kklobe** [comment]
Thanks @Burrito78! Just a slight loosening of the tolerance took care of it, I think:

CleanShot 2023-11-25 at 09 52 16@2x
CleanShot 2023-11-25 at 09 51 31@2x

**Comment by rderooy** [comment]
should this not be limited to affected architectures, or to non-x86_64?

File: src/fpu/fpu_instructions.h [hunk 1]
```diff
@@ -102,9 +103,29 @@ static void FPU_FPOP(void){
...
+		if (fpu.cw & ExtendedPrecisionModeMask) {
+			constexpr double tolerance = 1e-14;
+			const auto lower           = std::floor(in);
+			const auto upper           = std::ceil(in);
+
+			if ((in - lower) < tolerance) {
+				return lower;
+			} else if ((upper - in) < tolerance) {
+				return upper;
+			}
```
Comment (kcgen) [inline src/fpu/fpu_instructions.h:125]: (just a minor micro-logic comment):
- `upper` isn't needed until we fail the first `lower` check,
- and both `upper` and `lower` aren't needed once the if/else fails

So could be a nice use of C++17's if/else initializer (moving `lower` and `upper` in as initializers).

File: src/fpu/fpu_instructions.h [hunk 2]
```diff
@@ -419,13 +440,12 @@ static void FPU_FUCOM(Bitu st, Bitu other){
 }
 
 static void FPU_FRNDINT(void){
-	int64_t temp  = static_cast<int64_t>(FROUND(fpu.regs[TOP].d));
-	double tempd = static_cast<double>(temp);
+	const auto rounded  = FROUND(fpu.regs[TOP].d);
```
Comment (kcgen) [inline src/fpu/fpu_instructions.h:443]: Very nice discombobulating :-)

**Comment by Burrito78** [comment]
Confirmed to work now as expected! Man, this is really great! Thank you so much for doing this! 🥳

Bildschirmfoto 2023-11-25 um 20 43 07

**Comment by kklobe** [comment]
> should this not be limited to affected architectures, or to non-x86_64?

I think on x86_64 `fpu_instructions_x86.h` gets pulled in instead of this file, but I haven't done the full matrix testing on the different platforms yet - good reminder, thanks.

**Review by kcgen — APPROVED** [review]
Extremely nice work putting this together, @kklobe.
Looks good from my side!
Sooner we get this in, the sooner we can get wider test coverage.

**Comment by kklobe** [comment]
> Extremely nice work putting this together, @kklobe. Looks good from my side! Sooner we get this in, the sooner we can get wider test coverage.

Thanks! Just want to get in a skip for x64 platforms, as mentioned by @rderooy, and test the artifacts on Rosetta and Windows 11, shouldn't take long.

File: src/fpu/fpu_instructions.h [hunk 3]
```diff
@@ -102,9 +103,33 @@ static void FPU_FPOP(void){
...
-	case ROUND_Down: return (floor(in));
-	case ROUND_Up: return (ceil(in));
-	case ROUND_Chop: [[fallthrough]]; // cast by the caller chops
+	case ROUND_Down: return std::floor(in);
+	case ROUND_Up: return std::ceil(in);
+	case ROUND_Chop: 
+	{
+		// Most x32/x64 builds use fpu_instructions_x86.h.
+		// If we are here on x64, we don't need this fix.
+#if !defined(__x86_64__) && !defined(_M_X64)
```
Comment (kklobe) [inline src/fpu/fpu_instructions.h:112]: @kcgen let me know if you are ok with this here, or if you would prefer a new build system define get created and used instead.

**Comment by johnnovak** [comment]
> > Extremely nice work putting this together, @kklobe. Looks good from my side! Sooner we get this in, the sooner we can get wider test coverage.
> 
> Thanks! Just want to get in a skip for x64 platforms, as mentioned by @rderooy, and test the artifacts on Rosetta and Windows 11, shouldn't take long.

Hang on a minute, so as I said yesterday on the chat @kklobe, our 32-bit build works fine (it doesn't manifest the vertical stripes bug) probably because somehow the 80-bit FPU regs get used by the C libraries. So x86_64 is fine, but I think it's still brittle... (what it the compiler stops using the 80-bit FPU regs one day).

But the 64-bit builds, so x86_64, however are broken and manifest the bug 100% as 80-bit FPU regs are not used for 64-bit x86 builds. We absolutely want this fir for the x86_64 builds!

The only thing you could exclude theoretically is the 32-bit x86 builds... but I wouldn't bother.

**Comment by kklobe** [comment]
> > > Extremely nice work putting this together, @kklobe. Looks good from my side! Sooner we get this in, the sooner we can get wider test coverage.
> > 
> > 
> > Thanks! Just want to get in a skip for x64 platforms, as mentioned by @rderooy, and test the artifacts on Rosetta and Windows 11, shouldn't take long.
> 
> Hang on a minute, so as I said yesterday on the chat @kklobe, our 32-bit build works fine (it doesn't manifest the vertical stripes bug) probably because somehow the 80-bit FPU regs get used by the C libraries. So x86_64 is fine, but I think it's still brittle... (what it the compiler stops using the 80-bit FPU regs one day).
> 
> But the 64-bit builds, so x86_64, however are broken and manifest the bug 100% as 80-bit FPU regs are not used for 64-bit x86 builds. We absolutely want this fir for the x86_64 builds!
> 
> The only thing you could exclude theoretically is the 32-bit x86 builds... but I wouldn't bother.

Right, this is just for the rounding fix, which isn't needed on x64 builds, as shown by @Burrito78's screenshots for main on Rosetta above.

File: src/fpu/fpu_instructions.h (hunk 3 above)
Comment (kcgen) [reply src/fpu/fpu_instructions.h:112]: Good point - if the CPU type isn't commonly defined, Meson could genericize them for us.  Seeing them used in the DR codecs, XXHash, and STB Vorbis libs.. and these are extremely universal headers compiled by countless compilers; so I think these are defacto enough.

Comment (johnnovak) [reply src/fpu/fpu_instructions.h:112]: No, this can't be correct. As I explained below, the bug 100% manifest in our 64-bit x86 builds—I tried it yesterday on Win10. I also tried the 0.80.1 32-bit build, and the 32-bit x86 build is "fine" because that uses 80-bit FPU regs internally by fluke.

**Comment by johnnovak** [comment]
> Right, this is just for the rounding fix, which isn't needed on x64 builds, as shown by @Burrito78's screenshots for main on Rosetta above.

Ok... I slep like 5 hours 😬 Carry on, I was referring to the vertical stripes fix, so the other PR 😊

**Comment by Grounded0** [comment]
@johnnovak 

You're not the only one mixing these two up. At least now the PRs are correctly linked to the correct issues.

**Comment by kcgen** [comment]
@kklobe : maybe this rounding fix should be excluded for Intel x86 32bit machines, too? (__i386__, __i686__, ?) Hmm.. I'm not sure of what preprocessor macros are common for these.

File: src/fpu/fpu_instructions.h [hunk 4]
```diff
@@ -102,9 +103,33 @@ static void FPU_FPOP(void){
...
+			constexpr double tolerance = 1e-14;
+
+			if (const auto lower = std::floor(in);
+			    (in - lower) < tolerance) {
+				return lower;
+			} else if (const auto upper = std::ceil(in);
+			           (upper - in) < tolerance) {
+				return upper;
+			}
+		}
```
Comment (johnnovak) [inline src/fpu/fpu_instructions.h:129]: *Maybe* this could cause issues with application software that use heavy numerical calculations, but who knows... I guess only testing will tell!

**Comment by kklobe** [comment]
> @kklobe, should it be excluded for Intel x86 32bit machines, too? (`__i386__`, `__i686__`, ?) Hmm.. I'm not sure of what preprocessor macros are common for these.

Those will be skipped entirely, because the 32-bit x86 builds all use `fpu_instructions_x86.h` - MSVC hard-defines it (`C_FPU_X86` in `config.h`), and Meson defines it for both 32 and 64 bit x86 builds.

File: src/fpu/fpu_instructions.h (hunk 4 above)
Comment (kklobe) [reply src/fpu/fpu_instructions.h:129]: Hopefully it should be more accurate for most cases! Like in Quake, there's that small difference when rounding using truncate (or "chop") mode, that what is calculated as 8.0 on x64, on ARM is e.g. 7.999999999999996, but the chop mode puts that down to 7.

Comment (johnnovak) [reply src/fpu/fpu_instructions.h:129]: A disclaimer section in our webpage is in order, btw. Basically we should state our emulator is only for entertainment purposes and for personal use, and we're not responsible for any loss or catastrophes when used for anything else... 😬

Comment (MasterO2) [reply src/fpu/fpu_instructions.h:129]: > A disclaimer section on our webpage is in order, btw. Basically we should state our emulator is only for entertainment purposes and for personal use, and we're not responsible for any loss or catastrophes when used for anything else... 😬

And a sub-disclaimer for those engineers who, for some strange reason, are using Dosbox Staging to engineer bridges.... [/half joking]

Comment (ThomasEricB) [reply src/fpu/fpu_instructions.h:129]: > A disclaimer section on our webpage is in order, btw. Basically we should state our emulator is only for entertainment purposes and for personal use, and we're not responsible for any loss or catastrophes when used for anything else... 😬

Yeah we can do something like that.

Comment (kcgen) [reply src/fpu/fpu_instructions.h:129]: > *Maybe* this could cause issues with application software that use heavy numerical calculations, but who knows... I guess only testing will tell!

With rounding pumped up a notch, we should see improvements in truss thicknesses and I-beam width calculations, so perhaps safer buildings and bridges: 👍 👷

I wish I had some of those old posts.. but yeah: there are some very old civil engineering packages that people are still using and teaching with in schools (and using dosbox!)

Comment (Grounded0) [reply src/fpu/fpu_instructions.h:129]: I think there should be a setting where we default to this ON and optionally allowing people to run with more precision if they so desire (on ARM platforms).

Comment (Grounded0) [reply src/fpu/fpu_instructions.h:129]: But that can be implemented as a follow-up PR.

**Comment by kcgen** [comment]
> Those will be skipped entirely, because the 32-bit x86 builds all use fpu_instructions_x86.h

Thanks for clearly this up, @kklobe; perfect!

File: src/fpu/fpu_instructions.h (hunk 4 above)
Comment (kklobe) [reply src/fpu/fpu_instructions.h:129]: This is an awfully big hammer for one feature in one game...

Comment (Grounded0) [reply src/fpu/fpu_instructions.h:129]: Its Quake. 🙂

**Comment by kklobe** [comment]
Another FYI: Quake switches precision mode quite often:

```log
2023-11-27 14:51:16.937 | FPU rounding mode: single precision
2023-11-27 14:51:16.937 | FPU rounding mode: single precision
2023-11-27 14:51:16.937 | FPU rounding mode: single precision
2023-11-27 14:51:16.937 | FPU rounding mode: single precision
2023-11-27 14:51:16.937 | FPU rounding mode: single precision
2023-11-27 14:51:16.937 | FPU rounding mode: extended precision
2023-11-27 14:51:16.937 | FPU rounding mode: extended precision
2023-11-27 14:51:16.937 | FPU rounding mode: single precision
2023-11-27 14:51:16.937 | FPU rounding mode: extended precision
2023-11-27 14:51:16.937 | FPU rounding mode: single precision
2023-11-27 14:51:16.937 | FPU rounding mode: extended precision
2023-11-27 14:51:16.937 | FPU rounding mode: single precision
2023-11-27 14:51:16.937 | FPU rounding mode: extended precision
2023-11-27 14:51:16.938 | FPU rounding mode: single precision
2023-11-27 14:51:16.938 | FPU rounding mode: single precision
2023-11-27 14:51:16.938 | FPU rounding mode: single precision
2023-11-27 14:51:16.938 | FPU rounding mode: single precision
2023-11-27 14:51:16.938 | FPU rounding mode: single precision
2023-11-27 14:51:16.938 | FPU rounding mode: extended precision
2023-11-27 14:51:16.938 | FPU rounding mode: extended precision
2023-11-27 14:51:16.938 | FPU rounding mode: single precision
2023-11-27 14:51:16.938 | FPU rounding mode: extended precision
2023-11-27 14:51:16.938 | FPU rounding mode: single precision
2023-11-27 14:51:16.938 | FPU rounding mode: extended precision
2023-11-27 14:51:16.938 | FPU rounding mode: single precision
2023-11-27 14:51:16.938 | FPU rounding mode: extended precision
2023-11-27 14:51:16.939 | FPU rounding mode: single precision
2023-11-27 14:51:16.939 | FPU rounding mode: single precision
2023-11-27 14:51:16.939 | FPU rounding mode: single precision
2023-11-27 14:51:16.939 | FPU rounding mode: single precision
```

**Review by johnnovak — APPROVED** [review]
Looks good to me, let's merge then @kklobe 🚀

**Comment by kklobe** [comment]
> Looks good to me, let's merge then @kklobe 🚀

Wait! Critical fix incoming.
